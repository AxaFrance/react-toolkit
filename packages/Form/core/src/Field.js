import React from 'react';
import PropTypes from 'prop-types';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import FieldError from './FieldError';
import MessageTypes from './MessageTypes';
import FieldForm from './FieldForm';
import Constants from './FieldConstants';

const propTypes = {
  label: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  children: PropTypes.node.isRequired,
  message: PropTypes.string,
  messageType: PropTypes.oneOf([
    MessageTypes.error,
    MessageTypes.success,
    MessageTypes.warning,
  ]),
  isVisible: PropTypes.bool,
  forceDisplayMessage: PropTypes.bool,
  className: PropTypes.string,
  classModifier: PropTypes.string,
  id: PropTypes.string,
  classNameContainerLabel: PropTypes.string,
  classNameContainerInput: PropTypes.string,
};

const defaultClassName = 'row af-form__group';
const defaultProps = {
  message: null,
  messageType: MessageTypes.error,
  isVisible: true,
  forceDisplayMessage: false,
  className: defaultClassName,
  classModifier: '',
  id: '',
  classNameContainerLabel: Constants.defaultProps.classNameContainerLabel,
  classNameContainerInput: Constants.defaultProps.classNameContainerInput,
};

const Field = (props) => {
  const {
    children,
    message,
    label,
    messageType,
    forceDisplayMessage,
    classModifier,
    className,
    id,
    classNameContainerLabel,
    classNameContainerInput,
    isVisible,
  } = props;
  if (!isVisible) {
    return false;
  }

  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );

  return (
    <div className={componentClassName}>
      <div className={classNameContainerLabel}>
        <label className="af-form__group-label" htmlFor={id}>
          {label}
        </label>
      </div>
      <FieldForm
        className={classNameContainerInput}
        message={message}
        messageType={messageType}
        forceDisplayMessage={forceDisplayMessage}>
        {children}
        <FieldError message={message} messageType={messageType} />
      </FieldForm>
    </div>
  );
};

Field.propTypes = propTypes;
Field.defaultProps = defaultProps;

export default Field;
