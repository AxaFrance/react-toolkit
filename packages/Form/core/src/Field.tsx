import React, { ReactNode } from 'react';
import { ClassManager } from '@axa-fr/react-toolkit-core';
import FieldError from './FieldError';
import MessageTypes from './MessageTypes';

const defaultClassName = 'row af-form__group';

type FieldProps = {
  id: string;
  label: string;
  children?: ReactNode;
  message?: string;
  messageType?: MessageTypes;
  forceDisplayMessage?: boolean;
  className?: string;
  classModifier?: string;
  classNameContainerLabel?: string;
  classNameContainerInput?: string;
  isVisible?: boolean;
};

const defaultProps = {
  messageType: MessageTypes.error,
  forceDisplayMessage: false,
  className: defaultClassName,
  classNameContainerLabel: 'col-md-2',
  classNameContainerInput: 'col-md-10',
  isVisible: true,
};

const Field = ({
  label,
  children,
  message,
  messageType,
  forceDisplayMessage,
  classModifier,
  className,
  id,
  classNameContainerLabel,
  classNameContainerInput,
  isVisible,
}: FieldProps) => {
  if (!isVisible) {
    return <></>;
  }

  const componentClassName = ClassManager.getComponentClassName(
    className,
    classModifier,
    defaultClassName
  );

  return (
    <div className={componentClassName}>
      <div className={classNameContainerLabel}>
        <label className="af-form__group-label" htmlFor={id}>
          {label}
        </label>
      </div>
      <div className={classNameContainerInput}>
        {children}
        {forceDisplayMessage && (
          <FieldError message={message} messageType={messageType} />
        )}
      </div>
    </div>
  );
};

Field.defaultProps = defaultProps;

export default Field;
