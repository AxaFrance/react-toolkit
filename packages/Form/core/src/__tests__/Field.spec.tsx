import React from 'react';
import { render } from '@testing-library/react';
import Field from '../Field';

describe('<Field>', () => {
  it('renders Field correctly', () => {
    const { asFragment } = render(
      <Field id="id" label="label" message="message">
        <input type="text" />
      </Field>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
