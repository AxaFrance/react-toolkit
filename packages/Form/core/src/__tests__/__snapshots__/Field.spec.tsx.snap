// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Field> Render <Field/> when have children HelpMessage 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    >
      <small
        class="af-form__help"
      >
        Want help ?
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> when have children HelpMessage and have message and forceDisplayMessage equal true 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    >
      <small
        class="af-form__message af-form__message--error"
      >
        <span
          class="glyphicon glyphicon-exclamation-sign"
        />
        <span
          class="af-form__error-text"
        >
          error message
        </span>
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> when have children complete example with error message 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    >
      <div
        class="af-form__field af-form__field--error"
      >
        <input
          class="Ã¬nput--error"
          id="id"
          name="name"
          type="text"
        />
      </div>
      <small
        class="af-form__message af-form__message--error"
      >
        <span
          class="glyphicon glyphicon-exclamation-sign"
        />
        <span
          class="af-form__error-text"
        >
          error message
        </span>
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> when have children complete example without error message 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    >
      <div
        class="af-form__field"
      >
        <input
          id="id"
          name="name"
          type="text"
        />
      </div>
      <small
        class="af-form__help"
      >
        Want help ?
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> when have message and forceDisplayMessage equal true 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    >
      <small
        class="af-form__message af-form__message--error"
      >
        <span
          class="glyphicon glyphicon-exclamation-sign"
        />
        <span
          class="af-form__error-text"
        >
          error message
        </span>
      </small>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> when is not visible 1`] = `<DocumentFragment />`;

exports[`<Field> Render <Field/> when is visible 1`] = `
<DocumentFragment>
  <div
    class="row af-form__group"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    />
  </div>
</DocumentFragment>
`;

exports[`<Field> Render <Field/> with roleContainer and ariaLabelContainer 1`] = `
<DocumentFragment>
  <div
    aria-label="Label du groupe"
    class="row af-form__group"
    role="radiogroup"
  >
    <div
      class="col-md-2"
    >
      <label
        class="af-form__group-label"
        for=""
      >
        my label
      </label>
    </div>
    <div
      class="col-md-10"
    />
  </div>
</DocumentFragment>
`;
